clippy_lints/src/casts/fn_to_numeric_cast_any.rs

```diff
@@ -0,0 +1,34 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use rustc_errors::Applicability;
+use rustc_hir::Expr;
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self, Ty};
+
+use super::FN_TO_NUMERIC_CAST_ANY;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, cast_expr: &Expr<'_>, cast_from: Ty<'_>, cast_to: Ty<'_>) {
+    // We only want to check casts to `ty::Uint` or `ty::Int`
+    match cast_to.kind() {
+        ty::Uint(_) | ty::Int(..) => { /* continue on */ },
+        _ => return,
+    }
+
+    match cast_from.kind() {
+        ty::FnDef(..) | ty::FnPtr(_) => {
+            let mut applicability = Applicability::MaybeIncorrect;
+            let from_snippet = snippet_with_applicability(cx, cast_expr.span, "x", &mut applicability);
```

Comment:
You can avoid the snippet by using `span_lint_and_then` and suggest using `cast_expr.span.hi()` and `"()"`.

Reply:
Sorry I don't understand what you mean. I based this on the code in `fn_to_numeric_cast`

Reply:
Something like this

```rust
span_lint_and_then(
    cx,
    FN_CAST,
    cast_expr.span,
    &format!("casting function pointer `{}` to `{}`", from_snippet, cast_to),
    |diag| {
        diag.span_suggestion(
            cast_expr.span.hi(),
            "did you mean to invoke the function?",
            "()".to_string(),
            Applicability::MachineApplicable,
        );
    },
);
```

Reply:
Oh, I see, thanks!

The `from_snippet` is still used in the "casting function pointer `{}` to `{}`" message though, so I don't think it can be removed without altering that message.

I feel like the snippet adds something to both the error message and the help text, do you think we could merge as-is?

Reply:
The zero-sized span would show something like

    f() as usize
     ++

which makes the change a little more clear, but not a big deal.